pkg/resources/allocator.go

```diff
@@ -0,0 +1,112 @@
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package resources
+
+import (
+	"sort"
+
+	"github.com/golang/glog"
+
+	pluginapi "k8s.io/kubelet/pkg/apis/deviceplugin/v1beta1"
+
+	"github.com/k8snetworkplumbingwg/sriov-network-device-plugin/pkg/types"
+)
+
+// DeviceSet is to hold and manipulate a set of HostDevice
+type DeviceSet map[string]types.HostDevice
+
+// PackedAllocator implements the Allocator interface
+type PackedAllocator struct {
+}
+
+// NewPackedAllocator create instance of PackedAllocator
+func NewPackedAllocator() *PackedAllocator {
+	return &PackedAllocator{}
+}
+
+// NewDeviceSet is to create an empty DeviceSet
+func NewDeviceSet() DeviceSet {
+	set := make(DeviceSet)
+	return set
+}
+
+// Insert is to add a HostDevice in DeviceSet
+func (ds *DeviceSet) Insert(pciAddr string, device types.HostDevice) {
+	(*ds)[pciAddr] = device
+}
+
+// Delete is to delete a HostDevice in DeviceSet
+func (ds *DeviceSet) Delete(pciAddr string) {
+	delete(*ds, pciAddr)
+}
+
+// AsSortedStrings is to sort the DeviceSet and return the sorted keys
+func (ds *DeviceSet) AsSortedStrings() []string {
+	keys := make([]string, 0, len(*ds))
+	for k := range *ds {
+		keys = append(keys, k)
+	}
+	sort.Strings(keys)
+	return keys
+}
+
+// Allocate return the preferred allocation
+func (pa *PackedAllocator) Allocate(rqt *pluginapi.ContainerPreferredAllocationRequest, rp types.ResourcePool) []string {
+	size := rqt.AllocationSize
+	preferredDevices := make([]string, 0)
+
+	if size <= 0 {
+		glog.Warningf("Allocator(): requested number of devices are negative. requested: %d", size)
+		return []string{}
+	}
+
+	if len(rqt.AvailableDeviceIDs) < int(size) {
+		glog.Warningf("Allocator(): not enough number of devices were available. available: %d, requested: %d", len(rqt.AvailableDeviceIDs), size)
+		return []string{}
+	}
+
+	if len(rqt.MustIncludeDeviceIDs) > int(size) {
+		glog.Warningf("Allocator(): allocated number of devices exceeded the number of requested devices. allocated: %d, requested: %d",
+			len(rqt.MustIncludeDeviceIDs),
+			size)
+	}
+
+	availableSet := NewDeviceSet()
+	for _, available := range rqt.AvailableDeviceIDs {
+		dev, ok := rp.GetDevicePool()[available]
+		if ok {
+			availableSet.Insert(available, dev)
+		} else {
+			glog.Warningf("Allocator(): not available device id was specified: %s", available)
+			return []string{}
+		}
+	}
+	for _, required := range rqt.MustIncludeDeviceIDs {
+		_, ok := rp.GetDevicePool()[required]
+		if ok {
+			availableSet.Delete(required)
+		} else {
+			glog.Warningf("Allocator(): not available device was included: %s", required)
+			return []string{}
+		}
+	}
+	sortedAvailableSet := availableSet.AsSortedStrings()
```

Comment:
This will work if device to be allocated is a PCI device and device ID is a PCI device ID.

This  is not the case for AuxiliaryDevices of type SF (subfunction).
the ID/Name of such device is as follows: mlx5_core.sf.4  (4 being sort of device/instance index)

so allocator will not work as expected in this case unless the auxiliary device was created  in certain order.

to achieve correctness in such case, best approach is to have a separate allocator for auxiliaryNetDevices where it would perform the check of being allocated from the same parent PCI device. based on auxiliaryNetDevice pfAddr field.

(have buckets of auxiliaryNetDevice by their parent PCI device then allocate from the same bucket)

Alternative is to block creation of a resource pool if the type is AuxNetDevice and non default policy is provided.
Then deal with that as a followup.

Reply:
This is something that needs to brought up in our weekly community meetings to get an agreement on which direction its preferrable to go.

technically i think we can use the same allocator for all if implemented in a different way than whats proposed in this PR